### DISCUSSION TITLE
AAPS keeps saying recalculated value used

### DISCUSSION THREAD

**Discussion by TWilhoWI** [NightscoutFoundation/xDrip#3455] [Bug reports]
AAPS keeps saying "recalculated value used" with G7. I've followed the setup guide. The only thing i've enabled extra is "Special Pairing Workaround".
Version: 2024.04.19.
Phone: Cubot Mini 3

**Comment by Navid200** [comment]
This is a question for the AAPS team.  
But, I will tag someone who may be able to help.

@olorinmaia

**Comment by rpm5099** [comment]
Not to pile on in the wrong forum (sorry), I have noticed this as well @olorinmaia.

With G7 I've never seen it *not* show the recalculated data warning.  I think the warning is meant to be in response to noisy data, but I'm not sure.  This is also the case with the Libre 3, except the Libre 3 data is extremely smooth and virtually always 100% accurate, less noise than G6 and *way* less noise than G7.  I haven't looked at the code closely to figure out what it means but I have no idea what is actually being "recalculated".  It's also unclear which features do not function when this warning is present.

**Comment by TWilhoWI** [comment]
![1715375504831731650694512302192](https://github.com/NightscoutFoundation/xDrip/assets/120023149/7937b509-74c5-444d-a25f-5553e751623d)

Second shows missed reading?

**Reply by Navid200** [reply]
So, you have a connectivity issue in xDrip.  I wish you had said that from the beginning instead of bringing up something related to AAPS!

Is the disconnect expected, like when you have walked away from the phone?
Or, does it happen on a regular basis even when you are near the phone?

**Reply by TWilhoWI** [reply]
On regular basis.

**Reply by Navid200** [reply]
Does it unpair if you disable special pairing workaround?

**Reply by TWilhoWI** [reply]
For 5 minutes, but it reconnected back again

**Reply by Navid200** [reply]
That wasn't my question.  The only reason to enable special pairing workaround is if your phone unpairs on its own.  The special pairing workaround is a workaround of course.  That means that it is what we have to do because there is no other solution.  So, it should only be used, like any other workaround, when absolutely necessary.

If your phone does not unpair, you gain nothing by enabling special pairing workaround, but you may add negative side effects.

**Reply by Navid200** [reply]
If your phone unpairs, you will have to approve a pair request before it works again.  It will never reconnect on its own unless you enable special pairing workaround.

**Reply by TWilhoWI** [reply]
Hi Navid, my previous phone had this issue, that's why I enabled it.
Now trough the night my sensor status changed to "sensor errors"?

**Reply by TWilhoWI** [reply]
I had to re-enable to setting. The phone constantly disconnected.

**Reply by Navid200** [reply]
xDrip is free and you can use it as you wish.
We set it up such that it can be used with most phones, but may be not with all.
So, we explain how each setting should be used: https://navid200.github.io/xDrip/docs/Intermittent.html  

I am doing my best to help you.  I have tried to identify the word pair and clarify that you should not use the special pairing workaround if you phone does not unpair on its own.  You keep using the word disconnect.

We may have a language problem here and I am not going to insist.  I am sorry if it is not working for you.

**Reply by TWilhoWI** [reply]
Without the setting it is unpairing. So I guess I have to enable it. 
Now for the values which comes at different times? Is this because of the special pairing?

**Reply by Navid200** [reply]
Yes, if it unpairs, you should enable the special pairing workaround setting.

I need to do some experiments to understand the G7 behavior better to figure out the mechanism behind the additional transmissions.
Before I can do that or unless someone else can shed some light on this, I don't have enough knowledge to make any statements about them.

**Comment by olorinmaia** [comment]
This is not caused by xDrip, and is like Navid say an AAPS issue. Simply because G7 delivers BG data more often at irregular times compared to G6. AAPS expect BG data every 5 min, which doesn't cope well with G7 as each sensor got it's own transmitter and therefor BG interval will differ more often than G6. It also seem to send BG data at irregular times if a few readings are missed (but I'm a new user of G7 so not sure about this yet). I also notice that yellow warning sign in AAPS is activated almost all the time.

I recommend following this issue, and possibly commenting on how you consider this can be improved in AAPS:
https://github.com/nightscout/AndroidAPS/issues/3197

See here for more info about the yellow warning sign:
https://androidaps.readthedocs.io/en/latest/Getting-Started/Screenshots.html#yellow-warning-sign

**Reply by Navid200** [reply]
I'm not aware of G7 sending readings at irregular time intervals.  This is the first time I am hearing it.  
I cannot say that is an incorrect statement.  All I am saying is that I will be very surprised if that was the case.
But, at the same time, I have to say that as this is the first time I am hearing it, I cannot claim that I can prove otherwise.

My thought is that a G7 can be used with a pump and an app at the same time the same as G6.
The fact is that soon, it will be able to be used by also an Apple watch.  But, let's put that one on hold for now.
Considering the pump and the app, either one of those needs to wake up exactly in 5 minutes to be ready to capture the next reading.  
If G7 changes the time interval because one of the two had a problem, how could the other one synchronize?

I am looking at my data table, and all the readings I was able to look at scrolling back line up on one reading every 5 minutes interval.

I know that there are phones that disable apps that run in the background.  xDrip is such an app.  When xDrip cannot wake up exactly in 5 minutes, it will miss a reading.  In that case, on the next reading, it also backfills.  
If AAPS does not like occasional missing readings, this could be the cause.  But, I am not an AAPS user.  

I agree that every G7 has its own transmitter with a local time reference.  This is identical to every G6 transmitter.  So, if you did experience a particular problem for every new G6, you can expect to experience that same problem for every G7.  

Please feel free to correct me or give me info I may be missing.
Thanks

**Reply by olorinmaia** [reply]
Im not sure about this one either. But will provide an example if what i think happend happens again 😉

**Reply by Navid200** [reply]
Yes please.  Thanks

**Reply by olorinmaia** [reply]
It happend again. We hadn't received bg in 15 min. So I restarted g7 collector manually and it connected and fetched bg immediately around 10:25. Which is a irregular time and makes AAPS recalculate bg data arrival. This isn't a problem as I see it, rather positive that g7 try to send out more often when data is missed. 

I turned on attached setting afterwards on my sons phone, is that a good idea? 

![Screenshot_20240511_110906_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/114103483/92b1938b-5243-443c-b21e-bfb4419d3544)


![Screenshot_20240511_104040_xDrip+ variant1~2](https://github.com/NightscoutFoundation/xDrip/assets/114103483/36684023-8d11-49b1-b1cd-9d564a4891d3)

**Reply by Navid200** [reply]
Aggressive service restarts should be enabled for all phones.
I will add to the recommended settings to ensure no one disables it.  Thanks

Thanks for the data table.  I had not noticed that before.  Thanks for posting that.  
This brings up different possibilities considering we don't know what is happening.  I need to create a test xDrip to see if I can figure out what's going on.

The way I see it, either G7 transmits each analog reading twice.  Or, it takes into account the handshakes from a receiving partner and sends additional data when there are missed handshakes.
Either of those will have different issues.  That's why it is not clear to me what's going on.
Thanks so much for sharing this.

**Reply by TWilhoWI** [reply]
Navid, do you need extra data or logging to further investigate this?

**Reply by Navid200** [reply]
I think you are referring to the connectivity issue and not what the title of this thread is, right?

Our knowledge on what leads to best results for all phones is summarized here for intermittent connectivity:
https://navid200.github.io/xDrip/docs/Intermittent.html  

If you are not sure that you can follow it, you can post screenshots of your settings here and I will have a look.

There are a few settings only that can improve performance depending on the phone brand and model and we have stated that it may help to enable special pairing workaround or it may help if you disable minimize scanning or it may help if you enable it and also enable avoid scanning.

If you experience a change of behavior depending on xDrip version, please report it clearly.  Tell us which version of xDrip fails and which works.  Ideally, you identify two consecutive xDrip releases where only one causes the problem and not the other.

There are so many phone brands and models.  It is not practical for us to test them all.  It would be very expensive.  I personally only have 6 phones.  Neither is Cubot.  
Based on user reports, Pixel is the best.  Xiaomi is the worst.  I have seen a few problem reports with Cubot.

**Reply by TWilhoWI** [reply]
Hi,
I did follow the best practices guide. I will try with the setting minimize scanning disable to see if the disconnect are better?

**Reply by TWilhoWI** [reply]
@Navid200 I'm still getting the issue that Xdrip is missing a lot of the readings of the G7. What can I do? I've followed all the best practices.

**Reply by Navid200** [reply]
I am running some experiments to see if I can do anything to improve the G7 collector.
This is where I am including the details about my experiments in this regard: https://navid200.github.io/xDrip/docs/Dexcom/G7OnceAMinuteScan.html

At this point, this is just a thought.  I hope to be able to improve things.  Thanks for your patience.

**Reply by Navid200** [reply]
Can you also make sure that the following setting is enabled?
Settings -> Less common settings -> Aggressive service restarts

**Reply by TWilhoWI** [reply]
It is enabled

**Reply by Navid200** [reply]
I have opened a PR to improve this: https://github.com/NightscoutFoundation/xDrip/pull/3461
My hope is that it will not increase battery consumption.  I am testing it now.
We now have to be patient and wait for a review.

**Reply by TWilhoWI** [reply]
Thanks Navid! Please keep me posted!

**Comment by Navid200** [comment]
Please use the xDrip release of May 22, 2024 or any after that.
xDrip used to get confused and screw up the timetamps when G7 went into the once a minute transmission.  That does not happen any more.

So, you should not get the warning from AAPS any longer.
Of course, you may still get it every time you start a new G7.  But, you should never get it just because of a disconnect.

**Comment by TWilhoWI** [comment]
Hi @Navid200 , repost. I still have some disconnections.
![333574213-39210120-6123-4819-9825-d6cccd0b4751](https://github.com/NightscoutFoundation/xDrip/assets/120023149/7b16ba96-33f7-4e71-a411-0581dac325fa)
![333634116-abc052ec-6204-467d-a631-50cb26198973](https://github.com/NightscoutFoundation/xDrip/assets/120023149/d6b5d175-9151-4a12-89f5-ae186faec5db)

**Reply by Navid200** [reply]
This phone was paired, but unpaired on its own?
Or is this phone never paired yet?

**Reply by TWilhoWI** [reply]
It was unpaired at it's own. It was paired before. Also it did paired back at his own, but after the 15min mark.

**Reply by Navid200** [reply]
The following is going to sound like an excuse.  But, it is a fact.
This is too complicated for me to fix.  The person who could possibly fix it is extremely busy with everything going on right now including Android 15 being around the corner and we are still not ready for it!

By any chance, can you use a different phone while this gets looked at?
Any Pixel phone would have none of these connectivity issues.

**Reply by TWilhoWI** [reply]
Completely understandable.
Unfortunately I don't have any other phone to my disposal.
Any chance the special pairing workaround will work to overcome this issue?

**Reply by Navid200** [reply]
After a pair is established between a Dexcom device and your phone, it will not unpair.
But, Samsung uses non-standard features on their phones.

The special pairing workaround was added to address the mess caused by Samsung.

If a phone other than Samsung unpairs with allow OB1 unbinding setting disabled, we suggest enabling special pairing workaround and giving it a try to see if it may make a difference even though it is not a Samsung.

**Reply by TWilhoWI** [reply]
It's a Cubot mini 3 phone. Seeing this error above, will this classify as a unpairing?
I'm willing to give the special pairing workaround a test?

**Reply by Navid200** [reply]
Yes, it does since you approved a pair request and it is still needing to pair again, which means that it unpaired on its own.
Unless, you went to the phone Bluetooth settings and forgot the paired device.

**Reply by TWilhoWI** [reply]
Perfect, I will enable the setting. 
Does it require a reboot of the device?

**Reply by Navid200** [reply]
It wouldn't hurt

**Reply by TWilhoWI** [reply]
Using this setting it seems very stable now!
